### ISSUE TITLE
Unexpected console error occurs. When useQuery with same key fails on multiple child components with suspense.

### ISSUE THREAD

**Issue by grap3fruit** [tannerlinsley/react-query#3550]
### Describe the bug

I am trying to use Suspense in React 18.

However, if useQuery with the same key fails in multiple child components, there is a problem that an error log occurs a lot.

It's fine for different useQuery to fail. This problem occurs when useQuery with the same key fails.

In fact, in the service I'm working on, about 12300 errors were output in the same situation, and the browser was very slow. (This was a case where useQuery with the same key called from 3 components failed.)

As of now, the solution I've found is to change the parent component to only call useQuery once and pass the value as props to the child component. But I don't know if this is the best.

![image](https://user-images.githubusercontent.com/13213473/164887154-00b92afd-2764-4a50-82c2-284110b22fd5.png)

### Your minimal, reproducible example

https://codesandbox.io/s/mystifying-sun-jvt227?file=/pages/_app.jsx

### Steps to reproduce

1. Open codesandbox link.
2. Open browser devtool.
3. Refresh browser inside the codesandbox.
4. Look at the number of `create Error` logs in devtool.

### Expected behavior

The expected behavior is, if an error occurs, immediately throw an error and terminate the task.
Or, normal components that are not related to the failing useQuery do not output an error.

Because this makes it very slow for the user to reach the error handler and error page.

### How often does this bug happen?

Every time

### Platform

-OS : Mac
-Browser : chrome
-Version : 99.0.4844.84

### react-query version

v3.38.0

### TypeScript version

v4.5.5

### Additional context

Thank you for your help.

**Comment by TkDodo** [comment]
this is not what I am seeing in that codesandbox. I have forked it, without changes:

https://eqiz11.sse.codesandbox.io/

what I see is:

- an error, likely from nextjs: `Uncaught Error: The server could not finish this Suspense boundary, likely due to an error during server rendering. Switched to client rendering.`
- a bunch of `client-hook-6.js:1 Error: Request failed with status code 404` errors. This is expected due to retries I guess.
- I then see a white page because there is no `ErrorBoundary` anywhere.

I have basically converted your example to a CRA setup, and it works similar there:

https://codesandbox.io/s/suspese-errors-9uurue

However, I cannot reproduce the "infinite errors", not even in your sandbox. So I'm guessing the screenshot shown isn't from the sandbox?

**Comment by grap3fruit** [comment]
Oh yes sorry. You are right.

The 12305 errors in the screenshot are from my actual project, not from codesandbox. It wasn't exactly an infinity error as you said, it took 12305 logs and then went to the error boundary.

I did not add error boundaries to the codesandbox I made as an example. I'll add it if needed.

But thankfully the issue seems to have been reproduced in the codesandbox project you created.

**Comment by TkDodo** [comment]
> I added some more components from the project you recreated. I hope this helps.

It doesn't matter if you render 3 different components that use the same query with the same key, or render the same component 3 times. It just makes the example more difficult to understand!

I'm still trying to understand the issue. I think what I'm seeing is expected. The way error boundaries work is that react catches all of them, logs them (apparently) and then shows the nearest error boundary.

Here is a quick example where the `FetchWrong` component doesn't really fetch, but just throws an Error during render (which is essentially what react-query is also doing when using ErrorBoundaries): 

https://codesandbox.io/s/suspese-errors-forked-td48fg?file=/src/day/fetch-wrong.tsx

As you can see, we render the component 3 times, so we see 3 log statements:

<img width="556" alt="Screenshot 2022-04-24 at 08 25 36" src="https://user-images.githubusercontent.com/1021430/164959720-d1383881-2cd9-4f2d-9cbe-b2c4e2fe4bb1.png">

Is that not what you would expect? Or is this not the issue?

**Comment by grap3fruit** [comment]
Sorry for the late reply.
I did some more debugging on the problem situation and found that it wasn't a problem with react query.

I haven't figured out the exact cause yet, but in React 18.0.0, I found that react-dom keeps re-rendering the errored component. Because of this, react query kept showing error logs. It was confirmed that it works normally in React 17.0.2.

If I find useful information, I'll leave a comment.
Thanks for your time.
